### DISCUSSION TITLE
How to drop local cache for S3 manually?

### DISCUSSION THREAD

**Discussion by RinChanNOWWW** [ClickHouse/ClickHouse#61205] [Q&A] [answered]
How to drop local cache for S3 manually? I want to test the performance with and without local cache.

**Comment by RinChanNOWWW** [comment]
And I have another question. When I read the codes in src/Disks/ObjectStorages/Cached/CachedObjectStorage.cpp, I found there is no logic about loading cahce in `readObject`.

<img width="1069" alt="企业微信截图_1e523fc8-b275-4768-9ed0-a1fcdd6632df" src="https://github.com/ClickHouse/ClickHouse/assets/33975039/2df203a4-99b9-4ba0-8d09-323078d735ec">

I'm confused.

**Comment by yiyang-shao** [comment, marked as answer]
> How to drop local cache for S3 manually? I want to test the performance with and without local cache.

Hi, you may refer to the following docs to see if they help.
* https://clickhouse.com/docs/en/operations/caches
* https://clickhouse.com/docs/en/operations/storing-data#using-local-cache
* https://clickhouse.com/docs/en/sql-reference/statements/system#drop-filesystem-cache

**Reply by RinChanNOWWW** [reply]
Thank you! My first question is solved.

And I have another question. Is the policy `s3_cache` conflict with `s3_main`?

I changed my config to 

```xml
<storage_configuration>
        <disks>
        <s3_disk>
            <type>s3</type>
            ... s3 configuration ...
            <metadata_path>lib/clickhouse/disks/s3_disk/</metadata_path>
        </s3_disk>
        <s3_cache>
            <type>cache</type>
            <disk>s3_disk</disk>
            <path>lib/clickhouse/disks/s3_cache/</path>
            <max_size>10Gi</max_size>
        </s3_cache>
        </disks>
        <policies>
            <s3_main>
                <volumes>
                <main>
                    <disk>s3_disk</disk>
                </main>
                </volumes>
            </s3_main>
            <s3_cache>
                <volumes>
                    <main>
                        <disk>cache</disk>
                    </main>
                </volumes>
            </s3_cache>
        </policies>
    </storage_configuration>
```

ClickHouse threw an exception: 

Code: 479. DB::Exception: Unknown disk cache: Cannot attach table `system`.`trace_log` from metadata file {MyWorkspace}/lib/clickhouse/store/xxx/trace_log.sql from query ATTACH TABLE system.trace_log

**Reply by yiyang-shao** [reply]
I guess the error is due to 
```xml
<s3_cache>
    <volumes>
        <main>
            <disk>cache</disk>    // Here, need to change to <disk>s3_cache</disk>
        </main>
    </volumes>
</s3_cache>
```

**Reply by RinChanNOWWW** [reply]
Thank you! It works.

**Comment by RinChanNOWWW** [comment]
Actually, when I wrote data with disk cache disabled and read data with disk cache enabled (by changing config.xml), ClickHouse didn't generate any cache.

My config file is like:


```xml
<storage_configuration>
    <disks>
    <s3_disk>
        <type>s3</type>
        <endpoint>xxx</endpoint>
        <access_key_id>xxx</access_key_id>
        <secret_access_key>xxx</secret_access_key>
        <metadata_path>lib/clickhouse/disks/s3_disk/</metadata_path>
    </s3_disk>
    <s3_cache>
        <type>cache</type>
        <disk>s3_disk</disk>
        <path>lib/clickhouse/disks/s3_cache/</path>
        <max_size>10Gi</max_size>
    </s3_cache>
    </disks>
    <policies>
    <s3_main>
        <volumes>
        <main>
            <disk>s3_disk</disk>
        </main>
        </volumes>
    </s3_main>
    </policies>
</storage_configuration>
```

**Reply by yiyang-shao** [reply]
Are you able to get any result from `SHOW FILESYSTEM CACHES`

**Reply by RinChanNOWWW** [reply]
Yes. 

<img width="502" alt="image" src="https://github.com/ClickHouse/ClickHouse/assets/33975039/2586ef16-46e7-4570-bfd7-2d7927cfbf98">

But ClickHouse still doesn't download any objects into cache.

**Reply by yiyang-shao** [reply]
Then probably need to dig a bit deeper.
1. Make sure you also set the s3_cache under the `policies` session similar to the thread above.
2. Make sure all the settings are correctly configured and nothing disables the cache.
3. Try to run some SELECT queries.
4. DESCRIBE FILESYSTEM CACHE 's3_cache' and check `system. filesystem_cache` and `system. filesystem_cache_log` to see if it contains more information.

**Reply by RinChanNOWWW** [reply]
I think I figured out the problems. It seems that I set a wrong storage policy when `CREATE TABLE`. I'm not familiar with the storage policy of ClickHouse, and I'm going to learn more about it.

Thank you very much!

**Comment by RinChanNOWWW** [comment]
May I ask another question @yiyang-shao. How can I disable and enable the file cache of the same table?

I have already created a table with `storage_policy = 's3_main'` (file cache is disabled), and I changed it to `storage_policy = 's3_cache'` (try to enable file cache) by 

```sql
alter table the_table modify setting storage_policy = 's3_cache'; 
```

(ref: #12632)

and this SQL executed successfully. I can also see `storage_policy='s3_cache'` when `SHOW CREATE TABLE the_table`.

But it seems that the new policy doesn't take effect. I want to disable and enable the cache for the same table to compare performance. Is there a possible way to do this?

FYI, the config is:

```xml
<storage_configuration>
        <disks>
            <s3_disk>
                <type>s3</type>
				....
                <metadata_path>lib/clickhouse/disks/s3_disk/</metadata_path>
            </s3_disk>
            <s3_cache>
                <type>cache</type>
                <disk>s3_disk</disk>
                <path>s3_cache/</path>
                <max_size>10Gi</max_size>
            </s3_cache>
        </disks>
        <policies>
            <s3_main>
                <volumes>
                    <main>
                        <disk>s3_disk</disk>
                    </main>
                </volumes>
            </s3_main>
            <s3_cache>
                <volumes>
                    <main>
                        <disk>s3_cache</disk>
                    </main>
                </volumes>
            </s3_cache>
        </policies>
    </storage_configuration>
```

**Reply by yiyang-shao** [reply]
For simplicity, probably you can try `enable_filesystem_cache` setting for each query to compare the performance without modifying the storage policy.

Reference
> enable_filesystem_cache - allows to disable cache per query even if storage policy was configured with cache disk type. Default: true.

**Reply by RinChanNOWWW** [reply]
Thank you. 

And I found that the storage policy modification took effect after I relaunched ClickHouse.
